src/Data/Arena.hs

```diff
@@ -15,74 +17,169 @@ module Data.Arena (
     withUnmanagedArena,
 ) where
 
-import           Control.DeepSeq (NFData (..), rwhnf)
+import           Control.DeepSeq (NFData (..))
+import           Control.Exception (assert)
 import           Control.Monad.Primitive
+import           Data.Bits (complement, popCount, (.&.))
 import           Data.Primitive.ByteArray
 import           Data.Primitive.MutVar
+import           Data.Primitive.MVar
+import           Data.Primitive.PrimVar
 
 #ifdef NO_IGNORE_ASSERTS
-import           Control.Monad (forM_)
 import           Data.Word (Word8)
 #endif
 
-data ArenaManager s = ArenaManager
+data ArenaManager s = ArenaManager (MutVar s [Arena s])
 
 newArenaManager :: PrimMonad m => m (ArenaManager (PrimState m))
 newArenaManager = do
-    _toUseTheConstraint <- newMutVar 'x'
-    return ArenaManager
+    m <- newMutVar []
+    return $ ArenaManager m
 
 -- | For use in bencmark environments
 instance NFData (ArenaManager s) where
-    rnf _ = ()
+    rnf (ArenaManager !_) = ()
 
--- TODO: this is debug implementation
--- we retain all the allocated arrays,
--- so we can scramble them at the end.
-data Arena s = Arena (MutVar s [MutableByteArray s])
+data Arena s = Arena
+    { curr :: !(MVar s (Block s))   -- current block, also acts as a lock
+    , free :: !(MutVar s [Block s])
+    , full :: !(MutVar s [Block s])
+    }
+
+data Block s = Block !(PrimVar s Int) !(MutableByteArray s)
 
 instance NFData (Arena s) where
-  rnf (Arena mvar) = rwhnf mvar
+  rnf (Arena !_ !_ !_) = ()
 
 type Size      = Int
 type Offset    = Int
 type Alignment = Int
 
+blockSize :: Int
+blockSize = 0x100000
```

Comment:
Might be worth tuning, but seems like a good default for now. :+1:

Reply:
```
Prelude Numeric> showHex (4096 * 256) ""
"100000"
```

So it's close. For better test coverage it should be a lot less, for real use case it might sense to double.